### ISSUE TITLE
Feature Request: Disable/Reschedule maintenance upgrades which interrupt running Container Apps

### ISSUE THREAD

**Issue by gdellolio-mariner** [microsoft/azure-container-apps#1144]
**Is your feature request related to a problem? Please describe**.  

We currently have a single Container App running in Azure. We noticed that the application would randomly be gracefully stopped and restarted. This application is supposed to be running 24/7 and should be run uninterrupted. Due to the nature of the application, we can only have one instance (replica) of this container running. 

After more digging, we found that Container Apps unknowingly and occassionally does behind-the-scenes updates to Container Apps which automatically restarts your running containers. For those that are running a single container, it gracefully stops your container and bring it up on a different node causing a brief period of downtime. This is problematic when you need your container to be running indefinitely.

What makes matters worse - no one knows when these updates occur so it's impossible to plan around them. In addition to not knowing when the updates will happen (and when your containers will automatically restart on you), these updates seem to happen at any time of day... not just "outside business hours". This makes Container Apps almost useless for critical applications that rely on 100% uptime.

**Describe the solution you'd like**.    

A few solutions:
1. Users should be able to schedule a maintenance window that works for them. This will specify when it's ok to do updates in the background.
2. Background updates should be scheduled during "outside business hours" for the region in which it is running. 

**Describe alternatives you've considered**.   

I've considered using a different solution for hosting my container app.

1. AKS is a good workaround but would add way too much overhead for a simple single container application. 
2. ACI is another option, but apparently this same exact upgrade process happens there too.
4. Using a VM - This strips away all the benefits of using Azure to host applications.

**Additional context**.  

I understand the importance of regular upgrades to ensure stability and security.  I don't understand why we have to sacrifice reliability and uptime for stability and security. Users should, at the very least, be able to select the maintenance window in which these updates occur. If that's not a possibility, updates should be done on off hours based on the region.

**Comment by sczd** [comment]
I agree with @gdellolio-mariner we should be able to pick a maintenance window, it's a prerequisites to use this service in production.
Also, the underlying infrastructure upgrades should move replicas and revisions one by one on each availability zones, to eliminate downtime. I can't find any confirmation of this process in the documentation.

**Comment by dsczltch** [comment]
We agree with @gdellolio-mariner.
- First, we are not notified when upgrades happen. We did subscribe to Azure Service health alert and we are never notified when upgrade or maintenance run (we do for app service or sql database for example).
- Second, we cannot choose the maintenance window.
- Third, containers partially silently restart. We get log when they restart but metric "Replica Restart Count" do no show this restart, which make things inconsistent.

**Comment by vportascarta** [comment]
+1
I am in the same situation. I have opened a support ticket to have more info on this.

**Comment by mapperCoderZ** [comment]
Hello, did you find a workaround or a solution for this ? We have a similar restart, many times a month, we don't know how to fix it. Thank you in advance for your answers

**Comment by gdellolio-mariner** [comment]
unfortunately not and there hasn't been any acknowledgement on this issue either

**Comment by vportascarta** [comment]
In the response I received, they are actively studying the implementation of a planned maintenance for ACA, but no deadline has been communicated. So on my side, I will have to switch to AKS to have more control over this.

**Comment by anthonychu** [comment]
Yes we've started work on adding maintenance window support in Container Apps. Initially, the feature will enable you to pick a day of week, start time, and duration (e.g., Sunday, 02:00, 8 hours).

All times are UTC. Minimum window size is 8 hours. You can define multiple windows.

ETA for public preview is end of this calendar year or sooner.

If you have any feedback, please let us know.

**Comment by nlmoorreesj** [comment]
Also include daylight savings schedule please! We now had to do a nightly-shift just to update all the app-jobs times because of daylight savings. See also https://github.com/microsoft/azure-container-apps/issues/1109

**Comment by anthonychu** [comment]
@nlmoorreesj Unfortunately we're unable to support daylight saving time because not all of Container Apps' underlying infrastructure is able to support it. We may be able to support a UTC offset, but that's not going to help if you need adjustments for DST. After the feature reaches public preview, we'll investigate further.

**Comment by mikocot** [comment]
Being able to pick up the time is nice, but the most troubling part for me is thatcurrently it's almost impossible to confirm is maintanence was the reason for the apps being shut down.

That leads to unnecessary investigations and support tickets cositing both sides tons of time and money and makes ACA look simply unreliable.

Hence I propose to add a very clear entry in the `ContainerAppSystemLogs_CL` whenever maintenance is done and what is the scope.

All we get now is `RevisionUpdate` `Container app shutting down`, etc. with no apparent reason. What is even worse the support does not seem to be able to confirm if maintenance has happened or what. Transparent logging should solve all that..

**Comment by bettercalljason** [comment]
We have an ACA app that is especially mission-critical outside business hours (nighttime and weekends), so scheduled maintenance is desired _during_ business hours. Thus, we are very happy that this under development.

We are currently investigating which workload profile - consumption or dedicated - we should choose. Since (correct me if I‘m wrong) apps under the consumption profile run on a shared node, will the changing of maintenance window times be possible for the consumption environment or only for the dedicated?

**Comment by anthonychu** [comment]
> We have an ACA app that is especially mission-critical outside business hours (nighttime and weekends), so scheduled maintenance is desired _during_ business hours. Thus, we are very happy that this under development.
> 
> We are currently investigating which workload profile - consumption or dedicated - we should choose. Since (correct me if I‘m wrong) apps under the consumption profile run on a shared node, will the changing of maintenance window times be possible for the consumption environment or only for the dedicated?

@bettercalljason That's a good question and we'll provide more details when this feature is ready. But you are correct, because consumption workload profiles run on shared infrastructure, some types of upgrades can still occur outside the maintenance window. _Initially_, only dedicated workload profiles will fully support planned maintenance.

I want to be clear though that this limitation is intended to be temporary, and we'll continue work on the feature to add full support for consumption workload profiles.

**Comment by anthonychu** [comment]
@mikocot Thanks for the feedback on logs. We have some of this planned as part of the maintenance windows feature. Agree that there needs to be better visibility in general to why a replica is shut down or restarted, and we'll see what can be improved.

**Comment by nlmoorreesj** [comment]
> [@nlmoorreesj](https://github.com/nlmoorreesj) Unfortunately we're unable to support daylight saving time because not all of Container Apps' underlying infrastructure is able to support it. We may be able to support a UTC offset, but that's not going to help if you need adjustments for DST. After the feature reaches public preview, we'll investigate further.

Not sure if an UTC offset helps, if daylight savings comes into play. Nevertheless can you at least add time-zones to the schedule when container apps, start? To me this looks like something outside the infrastructure, but just to the scheduling mechanism.

**Comment by mapperCoderZ** [comment]
any update on this request ? We have dedicated workload profiles, will this feature be released ? Will it be added in the ARM template Microsoft.App/containerApps ?
